File: solutions/misc/frequently-asked-questions/task-scheduler.js
/**
 * Task Scheduler (LC 621) — Medium
 * Pattern: Greedy + Max-Heap + Cooldown Queue
 *
 * Time:  O(N log 26) = O(N) since at most 26 task types
 * Space: O(26) = O(1)
 *
 * 60-Second ID:
 *   Core DS: task frequencies → max-heap; cooldown → queue
 *   Core Op: minimize total time (optimize scheduling)
 *   Constraint: same task must be n intervals apart
 *   → Greedy: always execute most frequent available task
 *
 * Alternate O(1) formula: max(N, (maxFreq-1)*(n+1) + countOfMaxFreq)
 */

/**
 * Simulation with Max-Heap + cooldown queue
 * @param {string[]} tasks
 * @param {number} n - cooldown
 * @returns {number} minimum intervals
 */
function leastInterval(tasks, n) {
  const freq = new Array(26).fill(0);
  for (const t of tasks) freq[t.charCodeAt(0) - 65]++;

  // Max-heap (negate for min-heap simulation)
  // Using sorted array as priority queue (26 elements max — acceptable)
  let maxHeap = freq.filter(f => f > 0).sort((a, b) => b - a);

  let time = 0;
  const queue = []; // [remaining_count, available_at_time]

  while (maxHeap.length > 0 || queue.length > 0) {
    time++;

    if (maxHeap.length > 0) {
      const top = maxHeap.shift() - 1; // execute most frequent
      if (top > 0) queue.push([top, time + n]); // requeue after cooldown
    }
    // idle cycle otherwise (time still increments)

    // Release tasks whose cooldown has expired
    if (queue.length > 0 && queue[0][1] === time) {
      const [cnt] = queue.shift();
      maxHeap.push(cnt);
      maxHeap.sort((a, b) => b - a); // keep sorted
    }
  }

  return time;
}

/**
 * O(1) formula approach
 * @param {string[]} tasks
 * @param {number} n
 * @returns {number}
 */
function leastIntervalFormula(tasks, n) {
  const freq = new Array(26).fill(0);
  for (const t of tasks) freq[t.charCodeAt(0) - 65]++;

  const maxFreq = Math.max(...freq);
  const countOfMax = freq.filter(f => f === maxFreq).length;

  return Math.max(tasks.length, (maxFreq - 1) * (n + 1) + countOfMax);
}

// ─── Tests ───────────────────────────────────────────────────────────────────
console.log(leastInterval(["A","A","A","B","B","B"], 2));        // 8
console.log(leastInterval(["A","A","A","B","B","B"], 0));        // 6
console.log(leastInterval(["A","A","A","A","A","A","B","C","D","E","F","G"], 2)); // 16

console.log(leastIntervalFormula(["A","A","A","B","B","B"], 2)); // 8

module.exports = { leastInterval, leastIntervalFormula };

File: solutions/misc/frequently-asked-questions/task-scheduler.py
"""
Task Scheduler (LC 621) — Medium
Pattern: Greedy + Max-Heap + Cooldown Queue

Time:  O(N log 26) ≈ O(N)
Space: O(26) = O(1)

Also: O(1) formula — max(N, (maxFreq-1)*(n+1) + countOfMaxFreq)
"""

import heapq
from collections import Counter, deque
from typing import List


def least_interval(tasks: List[str], n: int) -> int:
    """Simulation with max-heap (negated) + cooldown deque."""
    # Count frequency of each task type
    freq = Counter(tasks)
    # Build max-heap using negated values (Python heapq is min-heap only)
    max_heap = [-cnt for cnt in freq.values()]
    heapq.heapify(max_heap)

    time = 0
    # Queue stores (remaining_count, time_when_available) for tasks in cooldown
    queue: deque = deque()

    while max_heap or queue:
        time += 1

        if max_heap:
            # Execute the most frequent task (pop from max-heap)
            remaining = heapq.heappop(max_heap) + 1  # +1 since values are negated
            if remaining < 0:  # If task still has remaining occurrences, put in cooldown
                queue.append((remaining, time + n))
        # else: idle cycle (no task available, time still increments)

        # Restore tasks whose cooldown has expired back to the heap
        if queue and queue[0][1] == time:
            heapq.heappush(max_heap, queue.popleft()[0])

    return time


def least_interval_formula(tasks: List[str], n: int) -> int:
    """O(1) math formula.
    Key insight: The schedule consists of (max_freq - 1) full cycles of (n + 1) slots each,
    plus the final partial cycle containing all tasks with max_freq.
    """
    freq = Counter(tasks)
    max_freq = max(freq.values())
    # Count how many task types have the maximum frequency
    count_of_max = sum(1 for v in freq.values() if v == max_freq)
    # Answer is max of total tasks (no idle needed) or formula-based calculation
    return max(len(tasks), (max_freq - 1) * (n + 1) + count_of_max)


# ── Tests ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print(least_interval(list("AAABBB"), 2))               # 8
    print(least_interval(list("AAABBB"), 0))               # 6
    print(least_interval(list("AAAAAABCDEFG"), 2))         # 16

    print(least_interval_formula(list("AAABBB"), 2))       # 8
